pkg/vmprovider/providers/vsphere2/vmlifecycle/bootstrap_cloudinit.go

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package vmlifecycle
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/vmware/govmomi/vim25/types"
+	"gopkg.in/yaml.v2"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha2"
+	"github.com/vmware-tanzu/vm-operator/pkg/context"
+	"github.com/vmware-tanzu/vm-operator/pkg/util"
+	"github.com/vmware-tanzu/vm-operator/pkg/vmprovider/providers/vsphere2/constants"
+	"github.com/vmware-tanzu/vm-operator/pkg/vmprovider/providers/vsphere2/internal"
+	"github.com/vmware-tanzu/vm-operator/pkg/vmprovider/providers/vsphere2/network"
+)
+
+type CloudInitMetadata struct {
+	InstanceID    string          `yaml:"instance-id,omitempty"`
+	LocalHostname string          `yaml:"local-hostname,omitempty"`
+	Hostname      string          `yaml:"hostname,omitempty"`
+	Network       network.Netplan `yaml:"network,omitempty"`
+	PublicKeys    string          `yaml:"public-keys,omitempty"`
+}
+
+func BootStrapCloudInit(
+	vmCtx context.VirtualMachineContextA2,
+	config *types.VirtualMachineConfigInfo,
+	cloudInitSpec *vmopv1.VirtualMachineBootstrapCloudInitSpec,
+	bsArgs *BootstrapArgs) (*types.VirtualMachineConfigSpec, *types.CustomizationSpec, error) {
+
+	netPlan, err := network.NetPlanCustomization(bsArgs.NetworkResults)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create NetPlan customization: %w", err)
+	}
+
+	sshPublicKeys := bsArgs.BootstrapData.Data["ssh-public-keys"]
+	if len(cloudInitSpec.SSHAuthorizedKeys) > 0 {
+		sshPublicKeys = strings.Join(cloudInitSpec.SSHAuthorizedKeys, "\n")
+	}
+
+	metadata, err := GetCloudInitMetadata(string(vmCtx.VM.UID), bsArgs.Hostname, netPlan, sshPublicKeys)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var userdata string
+	if cloudInitSpec.RawCloudConfig.Name != "" {
+		// Check for the 'user-data' key as per official contract and API documentation.
+		// Additionally, to support the cluster bootstrap data supplied by CAPBK's secret,
+		// we check for a 'value' key when 'user-data' is not supplied.
+		// The 'value' key lookup will eventually be deprecated.
+		for _, k := range []string{cloudInitSpec.RawCloudConfig.Key, "user-data", "value"} {
+			if k != "" {
+				userdata = bsArgs.BootstrapData.Data[k]
+				if userdata != "" {
+					break
+				}
+			}
+		}
+
+		// NOTE: The old code didn't error out if userdata wasn't found, so keep going.
+
+	} else {
+		return nil, nil, fmt.Errorf("TODO: inlined CloudConfig")
```

Comment:
How do plan on supporting this?. 
do we need a JIRA/issue to track this somewhere?.

Reply:
Yes there is a series of lower priority BL items for these. My immediate goal is to get v1a2 to v1a1 parity, and then worry more about new functionality like the inline bootstrap configs here.